classes/question_ui_renderer.php

```diff
@@ -38,224 +37,100 @@
-        $xpath->registerNamespace("xhtml", self::XHTML_NAMESPACE);
-        $xpath->registerNamespace("qpy", self::QPY_NAMESPACE);
-
         $nextseed = mt_rand();
-        if ($qa->get_database_id() === null) {
+        $id = $this->attempt->get_database_id();
+        if ($id === null) {
             throw new coding_exception("question_attempt does not have an id");
         }
-        mt_srand($qa->get_database_id());
+
+        mt_srand($id);
         try {
-            $this->resolve_placeholders($xpath);
-            $this->hide_unwanted_feedback($xpath, $options);
-            $this->hide_if_role($xpath, $options);
-            $this->set_input_values_and_readonly($xpath, $qa, $options);
-            $this->soften_validation($xpath);
-            $this->defuse_buttons($xpath);
-            $this->shuffle_contents($xpath);
-            $this->add_styles($xpath);
-            $this->format_floats($xpath);
-            $this->mangle_ids_and_names($xpath, $qa);
-            $this->clean_up($xpath);
+            $this->resolve_placeholders();
+            $this->hide_unwanted_feedback();
+            $this->hide_if_role();
+            $this->set_input_values_and_readonly();
+            $this->soften_validation();
+            $this->defuse_buttons();
+            $this->shuffle_contents();
+            $this->add_styles();
+            $this->format_floats();
+            $this->mangle_ids_and_names();
+            $this->clean_up();
         } finally {
             // I'm not sure whether it is strictly necessary to reset the PRNG seed here, but it feels safer.
             // Resetting it to its original state would be ideal, but that doesn't seem to be possible.
             mt_srand($nextseed);
         }
 
-        return $newdoc->saveHTML();
+        $this->html = $this->xml->saveHTML();
+        return $this->html;
     }
 
     /**
      * Hides elements marked with `qpy:feedback` if the type of feedback is disabled in {@see question_display_options}.
      *
-     * @param DOMXPath $xpath
-     * @param question_display_options $options
      * @return void
      */
-    private function hide_unwanted_feedback(\DOMXPath $xpath, question_display_options $options): void {
+    private function hide_unwanted_feedback(): void {
         /** @var DOMElement $element */
-        foreach (iterator_to_array($xpath->query("//*[@qpy:feedback]")) as $element) {
-            $feedback = $element->getAttributeNS(self::QPY_NAMESPACE, "feedback");
+        foreach (iterator_to_array($this->xpath->query("//*[@qpy:feedback]")) as $element) {
+            $feedback = $element->getAttributeNS(constants::NAMESPACE_QPY, "feedback");
 
             if (
-                ($feedback == "general" && !$options->generalfeedback)
-                || ($feedback == "specific" && !$options->feedback)
+                ($feedback == "general" && !$this->options->generalfeedback)
+                || ($feedback == "specific" && !$this->options->feedback)
```

Comment:
Was passiert, wenn hier ein ungültiger Wert steht (weder "general" noch "specific")? Sollte dann nicht auch der Node entfernt werden, sicherheitshalber? Oder ein Fehler geworfen werden?

Reply:
Wenn dort ein ungültiger Wert steht, wird der Node ausgegeben. Ich fände es auch besser, wenn dieser einfach gelöscht wird.
Einen Fehler zu schmeißen, sehe ich da eher beim SDK.

Reply:
Ich schließe mich Jan an. Das ermöglicht auch eine bessere Vorwärtskompatibilität, wenn eine neue Feedback-Art eingeführt werden sollte. Ich habe das auch mal in Tefen dokumentiert und dieses Issue erstellt: https://github.com/questionpy-org/questionpy-sdk/issues/118